druid/src/text/layout.rs

```diff
@@ -0,0 +1,257 @@
+// Copyright 2020 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A type for laying out, drawing, and interacting with text.
+
+use std::ops::Range;
+
+use crate::kurbo::{Line, Point, Rect, Size};
+use crate::piet::{
+    Color, PietText, PietTextLayout, Text as _, TextAttribute, TextLayout as _,
+    TextLayoutBuilder as _,
+};
+use crate::{Data, Env, FontDescriptor, KeyOrValue, PaintCtx, RenderContext};
+
+/// Some text that can be drawn on the screen.
+///
+/// This wraps an inner layout object, and handles invalidating and rebuilding
+/// it as required.
+///
+/// This object is not valid until the [`rebuild_if_needed`] method has been
+/// called. Additionally, this method must be called anytime the text or
+/// other properties have changed, or if any  items in the [`Env`] that are
+/// referenced in this layout change.
+///
+/// In general, you should just call this method as part of your widget's
```

Comment:
"this method" should probably be "the `rebuild_if_needed` method"? Seems like maybe class and method got mixed up in cut'n'paste?